Source/WebKit/UIProcess/WebProcessProxy.cpp

```diff
@@ -1054,9 +1055,45 @@ void WebProcessProxy::gpuProcessExited(ProcessTerminationReason reason)
 #endif
 
 #if ENABLE(MODEL_PROCESS)
-void WebProcessProxy::createModelProcessConnection(IPC::Connection::Handle&&, WebKit::ModelProcessConnectionParameters&&)
+void WebProcessProxy::createModelProcessConnection(IPC::Connection::Handle&& connectionIdentifier, WebKit::ModelProcessConnectionParameters&& parameters)
 {
+    bool anyPageHasModelProcessEnabled = false;
+    for (auto& page : m_pageMap.values())
+        anyPageHasModelProcessEnabled |= page->preferences().modelProcessEnabled();
+    MESSAGE_CHECK(anyPageHasModelProcessEnabled);
+
+#if ENABLE(IPC_TESTING_API)
+    parameters.ignoreInvalidMessageForTesting = ignoreInvalidMessageForTesting();
+#endif
+
+#if HAVE(AUDIT_TOKEN)
+    parameters.presentingApplicationAuditToken = m_processPool->configuration().presentingApplicationProcessToken();
+#endif
+
+    ensureModelProcess().createModelProcessConnection(*this, WTFMove(connectionIdentifier), WTFMove(parameters));
 }
+
+void WebProcessProxy::modelProcessDidFinishLaunching(ProcessID)
+{
+    for (auto& page : m_pageMap.values())
+        page->modelProcessDidFinishLaunching();
+}
+
+void WebProcessProxy::modelProcessExited(ProcessID pid, ProcessTerminationReason reason)
+{
+    WEBPROCESSPROXY_RELEASE_LOG_ERROR(Process, "modelProcessExited: pid=%d reason=%{public}s", pid, processTerminationReasonToString(reason));
+
+    for (auto& page : m_pageMap.values())
+        page->modelProcessExited(reason);
+}
+
+ModelProcessProxy& WebProcessProxy::ensureModelProcess()
+{
+    if (!m_modelProcess)
+        m_modelProcess = ModelProcessProxy::create(*this);
```

Comment:
This doesn't seem right.  I don't think it's a good idea to have the WebProcessProxy owning and creating the ModelProcessProxy.  I imagine its lifetime will be managed more like the GPUProcessProxy.  Otherwise we are going to have an unnecessarily large number of model processes.